src/extension.ts

```diff
@@ -57,13 +101,39 @@ async function fetchPackageMetadata(requirement: ProjectNameRequirement): Promis
     return await metadataCache.get(requirement.name)!()
 }
 
+let lastLine = -1
+let lastHover: vscode.Hover | null = null
+let lastLineText = ''
+
 class PyPIHoverProvider implements vscode.HoverProvider {
     async provideHover(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.Hover | null> {
-        const requirement = parsePipRequirementsLineLoosely(document.lineAt(position.line).text)
+        let lineText = document.lineAt(position.line).text
+        if (lastLine === position.line && lineText === lastLineText) {
+            return lastHover
+        }
+
+        const isPyProjectToml =
+            document.languageId === 'toml' && document.fileName.toLowerCase().endsWith('pyproject.toml')
+
+        if (isPyProjectToml) {
+            const dependencies = getPyProjectDependencies(document.getText())
+            // Workaround for extracting just the package name from the line
+            lineText = lineText.split('=')[0].trim()
+            const requirement = dependencies.find((dependency) => dependency.name === lineText)
+            if (requirement === undefined) return null
```

Comment:
The only blocker: This doesn't actually work, because the `requirement` defined here goes out of scope as soon as the `if (isPyProjectToml)` branch ends two lines below – it's never actually used. This is an easy mistake to make while handling two different requirement extraction branches in one method, hence it'd be _really_ beneficial to refactor this the way I suggested in the previous comment

Reply:
Ah, actually there is a point in waiting to merge – this blocker is not actually addressed by your changes @alex-way! Because we're defining `const requirement` twice in the `provideHover`, we're parsing `pyproject.toml` lines twice, each time a different way, which is just confusing and error-prone. Do you see what I mean?